Update overview.md

mattiasvillani opened the pull request:
Rewrote parts of the text to make it more clear and streamlined.
See if you like it.

### PR Checklist

- [ ] Tests are added
- [ ] Entry in NEWS.md
- [x] Documentation, if applicable
- [ ] API changes require approval from a committer (different from the author, if applicable)

mattiasvillani committed:
Update overview.md

Rewrote parts of the text to make it more clear and streamlined.
See if you like it.

ToucheSir commented on docs/src/models/overview.md:
```diff
@@ -4,21 +4,19 @@ Flux is a pure Julia ML stack that allows you to build predictive models. Here a
 julia> using Flux
 
-julia> actual(x) = 4x + 2
+julia> actual(x) = 4x  + 2
```
Looks like a superfluous space was added here?

mattiasvillani replied:
Fixed.

ToucheSir commented on docs/src/models/overview.md:
````diff
@@ -29,7 +27,7 @@ This example will build a model to approximate the `actual` function.
 
 ```julia
-julia> x_train, x_test = hcat(0:5...), hcat(6:10...)
+julia> x_train, x_test = [0 1 2 3 4 5], [6 7 8 9 10]
````
Is this change necessary if the output line below already shows the expanded values?

mattiasvillani replied:
The hcat() construction was unnecessary, I think. Just takes the focus away from the essential for a person that is relatively new to Julia.

ToucheSir replied:
In that case, I think this line should have a semicolon so that the exact same thing isn't printed again below.

darsnack replied:
Let's do the semicolon. I agree the `hcat` is unnecessary.

mattiasvillani replied:
I agree with the semicolon. I am new to doing a PR on Docs with this many changes. Can I add the semicolon here here on the web interface on Github and commit, if so how?

ToucheSir commented on docs/src/models/overview.md:
```diff
@@ -48,44 +46,40 @@ Dense(1, 1)
 julia> model.weight
 1×1 Matrix{Float32}:
- -1.4925033
+ -0.067352235
```
There are a couple lines like this where the random outputs have changed. Since these aren't doctests, I don't see a reason for changing them (and thus making a noisier diff)

mattiasvillani replied:
The numbers in the original version do not seem to be consistent with each other. I think this is because Dense(1,1) is called twice as a result of the change from model() to predict(). Anyway, my numbers comes from a consistent run.

ToucheSir replied:
My concern is that this will make blames and anything else that uses the diff much harder/less reliable. Users already understand that randomness is involved, and adding maintenance burden for something purely cosmetic is a net negative in my books.

darsnack replied:
I think the only inconsistency is that `model` doesn't have the same values as `predict` (which is to be expected). All references to `predict`'s parameters seem consistent to me. If we are going with [this change](https://github.com/FluxML/Flux.jl/pull/1687#discussion_r681902194) then `model` won't exist, and most of these random number changes can be reverted.

mattiasvillani replied:
I believe that the predictions later are done with the parameters initialized by model, not by predict. Anyway, let me change so that we only use model (and make a note that predictions are generated by calling model on data, as mentioned in another comment.) and then revert to the first set of parameters from model.

ToucheSir commented on docs/src/models/overview.md:
````diff
@@ -48,44 +46,40 @@ Dense(1, 1)
-```julia
-julia> predict = Dense(1, 1)
-```
+A `Dense` layer is a struct with fields `weight` and `bias`. `weight` is a matrix of model weights and `bias` represents a bias vector. These model parameters are initialized randomly here, and will later be learned from data by optimization. 
````
```suggestion
Under the hood, a `Dense` layer is a struct with fields `weight` and `bias`. `weight` is a matrix of model weights and `bias` represents a bias vector. By default, these model parameters are initialized randomly.
```
"under the hood" is still useful because it's totally possible to implement a dense linear layer without a struct like this. It also gives a nod to the actual implementation in Flux (i.e. Flux is not magic).

mattiasvillani replied:
I prefer more direct writing and 'under the hood' as the phrase has already been used in the text and does not add much here.

ToucheSir commented on docs/src/models/overview.md:
```diff
@@ -48,44 +46,40 @@ Dense(1, 1)
+A `Dense` layer is a struct with fields `weight` and `bias`. `weight` is a matrix of model weights and `bias` represents a bias vector. These model parameters are initialized randomly here, and will later be learned from data by optimization. 
 
-`Dense(1, 1)` also implements the function `σ(Wx+b)` where `W` and `b` are the weights and biases. `σ` is an activation function (more on activations later). Our model has one weight and one bias, but typical models will have many more. Think of weights and biases as knobs and levers Flux can use to tune predictions. Activation functions are transformations that tailor models to your needs. 
+`Dense(1, 1)` also implements the function `σ(Wx+b)` where `W` and `b` are the weights and biases. `σ` is an activation function that makes the mapping from inputs to output nonlinear. Our model has one weight and one bias, but typical models will have many more. Think of weights and biases as knobs and levers Flux can use to tune predictions.
```
```suggestion
`Dense(1, 1)` implements the function `σ(Wx+b)` where `W` and `b` are the weights and biases. `σ` is an activation function that applies a (possibly nonlinear) transformation. Our model has one weight and one bias, but typical models will have many more. Think of weights and biases as knobs and levers Flux can use to tune predictions.
```
`σ`  is not necessarily nonlinear (and if not for compat reasons it might behoove us to rename it).

ToucheSir commented on docs/src/models/overview.md:
```diff
@@ -48,44 +46,40 @@ Dense(1, 1)
+julia> model(x_train)
 1×6 Matrix{Float32}:
- 0.0  -1.4925  -2.98501  -4.47751  -5.97001  -7.46252
+ 0.0  -0.0673522  -0.134704  -0.202057  -0.269409  -0.336761
```
See comment above about changing random values.

mattiasvillani replied:
See my previous reply. I would like to keep my numbers.

ToucheSir commented on docs/src/models/overview.md:
```diff
@@ -48,44 +46,40 @@ Dense(1, 1)
+More accurate predictions will yield a lower loss. This loss function is called [mean squared error](https://www.statisticshowto.com/probability-and-statistics/statistics-definitions/mean-squared-error/). You can write your own loss functions or rely on those already provided by Flux. Flux works by iteratively reducing the loss through *training*.
 
-Under the hood, the Flux [`train!`](@ref) function uses *a loss function* and *training data* to improve the *parameters* of your model based on a pluggable [`optimiser`](../training/optimisers.md):
+The Flux [`train!`](@ref) function uses *a loss function* and *training data* to improve the *parameters* of your model based on an [`optimiser`](../training/optimisers.md). We use the gradient `Descent` optimiser:
```
```suggestion
The Flux [`train!`](@ref) function uses *a loss function* and *training data* to improve the *parameters* of your model based on an [`optimiser`](../training/optimisers.md). Here, we use basic gradient descent:
```

ToucheSir commented on docs/src/models/overview.md:
```diff
@@ -98,59 +92,33 @@ julia> data = [(x_train, y_train)]
-
-The first parameter is the weight and the second is the bias. Flux will adjust predictions by iteratively changing these parameters according to the optimizer.
-
-This optimiser implements the classic gradient descent strategy. Now improve the parameters of the model with a call to [`train!`](@ref) like this:
+The [`train!`](@ref) function in Flux will use the optimizer to iteratively adjust the weights and biases in `parameters` to minimize the loss function. Calling `train!` once makes the optimizer take a single step toward the minimum: 
```
I would revert most of these changes wholesale. Tweak-wise, you could rename `predict` to `model` as you've done above and drop the lines showing the weight and bias again (the sentences on both sides will need to be joined).

mattiasvillani replied:
Not sure that I follow what you mean. Can you please clarify?

ToucheSir commented on docs/src/models/overview.md:
```diff
@@ -98,59 +92,33 @@ julia> data = [(x_train, y_train)]
-julia> parameters = params(predict)
-Params([[-0.99009055], [0.0]])
+julia> parameters = params(model)
+Params([[-0.067352235], [0.0]])
```
Another note about different random values.

mattiasvillani replied:
See my reply above.

ToucheSir commented on docs/src/models/overview.md:
````diff
@@ -98,59 +92,33 @@ julia> data = [(x_train, y_train)]
 ```
 julia> loss(x_train, y_train)
-267.8037f0
+186.32362365722656
````
Another note about different random values.

mattiasvillani replied:
See my reply above.

ToucheSir commented on docs/src/models/overview.md:
````diff
@@ -98,59 +92,33 @@ julia> data = [(x_train, y_train)]
 ```
 julia> parameters
-Params([[9.158408791666668], [2.895045275]])
+Params([[8.389461], [2.4336762]])
````
Another note about different random values.

mattiasvillani replied:
See my reply above.

ToucheSir commented on docs/src/models/overview.md:
````diff
@@ -98,59 +92,33 @@ julia> data = [(x_train, y_train)]
 ```
 
-It went down. Why? 
+as the parameters were changed by the algorithm:
````
```suggestion
And the parameters were changed by the optimizer:
```

ToucheSir commented on docs/src/models/overview.md:
```diff
@@ -98,59 +92,33 @@ julia> data = [(x_train, y_train)]
 
-The parameters have changed. This single step is the essence of machine learning.
-
-## Iteratively Train the Model
```
Why was this heading removed? I don't see a new heading to replace it.

mattiasvillani replied:
I don't think it is needed in the new version. The new heading 'Learn Model Parameters by Optimisation' captures this.

ToucheSir replied:
It is because the next sentence starts with "in the previous section..." :stuck_out_tongue:. My point was that there either needs to be a header breaking that up from the previous paragraph, or the sentence itself needs to be changed. Personally, I think the header captures the following text pretty well and don't see the motivation behind removing it.

ToucheSir commented on docs/src/models/overview.md:
````diff
@@ -98,59 +92,33 @@ julia> data = [(x_train, y_train)]
+Params([[8.389461], [2.4336762]])
 ```
 
-The parameters have changed. This single step is the essence of machine learning.
````
We could still keep around the second sentence here. I'm not tied to it though.

mattiasvillani replied:
I would not say that a single gradient descent step is the essence of ML, although I understand the intention.

ToucheSir commented on docs/src/models/overview.md:
```diff
@@ -159,36 +127,35 @@ julia> for epoch in 1:200
 
 julia> loss(x_train, y_train)
-0.007433314787010791
+0.0054461159743368626
```
Another note about different random values.

ToucheSir commented on docs/src/models/overview.md:
```diff
@@ -159,36 +127,35 @@ julia> for epoch in 1:200
 
 julia> parameters
-Params([[3.9735880692372345], [1.9925541368157165]])
+Params([[4.022609], [2.0063674]])
```
Another note about different random values.

mattiasvillani replied:
See my reply above.

ToucheSir commented on docs/src/models/overview.md:
```diff
@@ -159,36 +127,35 @@ julia> for epoch in 1:200
+julia> model(x_test)
 1×5 Array{Float64,2}:
- 25.8442  29.8194  33.7946  37.7698  41.745
+ 26.142  30.1646  34.1872  38.2099  42.2325
```
Another note about different random values.

mattiasvillani replied:
See my reply above.

ToucheSir commented on docs/src/models/overview.md:
```diff
@@ -159,36 +127,35 @@ julia> for epoch in 1:200
-First, we gathered real-world data into the variables `x_train`, `y_train`, `x_test`, and `y_test`. The `x_*` data defines inputs, and the `y_*` data defines outputs. The `*_train` data is for training the model, and the `*_test` data is for verifying the model. Our data was based on the function `4x + 2`.
-
-Then, we built a single input, single output predictive model, `predict = Dense(1, 1)`. The initial predictions weren't accurate, because we had not trained the model yet.
+As expected, the predictions are very accurate since the underlying function was simple and observed without noise. Later on we will see `Flux` in action on substantially more challenging applications.
```
Going to defer to other folks on whether the summary should stick around.

mattiasvillani replied:
I thought that it was too detailed.

darsnack replied:
I would keep this summary. It is intentionally overly detailed for absolute beginners.

ToucheSir commented on docs/src/models/overview.md:
```diff
@@ -159,36 +127,35 @@ julia> for epoch in 1:200
+4. Evaluate the learned model by assessing predictive performance on the test data.
 
-This overall flow represents how Flux works. Let's drill down a bit to understand what's going on inside the individual layers of Flux.
+Let's drill down a bit to understand what's going on inside the individual layers of Flux.
```
```suggestion
Let's drill down a bit to understand what's going on inside individual Flux layers.
```

ToucheSir commented on docs/src/models/overview.md:
```diff
@@ -159,36 +127,35 @@ julia> for epoch in 1:200
+As expected, the predictions are very accurate since the underlying function was simple and observed without noise. Later on we will see `Flux` in action on substantially more challenging applications.
 
-After building the model, we trained it with `train!(loss, parameters, data, opt)`. The loss function is first, followed by the `parameters` holding the weights and biases of the model, the training data, and the `Descent` optimizer provided by Flux. We ran the training step once, and observed that the parameters changed and the loss went down. Then, we ran the `train!` many times to finish the t ...
+In summary, Flux builds and evaluates a predictive model through the following steps:
```
```suggestion
In summary, building and evaluating a predictive model in Flux follows the following steps:
```
I think this reads a bit more accurately. Specifically, Flux itself doesn't do much in the way of evaluation, but rather provides (some) utilities for the user to do so.

ToucheSir commented on docs/src/models/overview.md:
```diff
@@ -159,36 +127,35 @@ julia> for epoch in 1:200
 
-After we trained the model, we verified it with the test data to verify the results. 
+1. Setup training and test data.
+2. Build a predictive model with initialized parameters.
```
```suggestion
2. Build a model with initialized parameters.
```
Redundant with the "summary" line above.

ToucheSir commented on docs/src/models/overview.md:
```diff
@@ -159,36 +127,35 @@ julia> for epoch in 1:200
+1. Setup training and test data.
+2. Build a predictive model with initialized parameters.
+3. Find better parameters by optimizing a loss function that penalizes poor predictions.
+4. Evaluate the learned model by assessing predictive performance on the test data.
```
```suggestion
4. Evaluate the learned model by assessing performance on test data.
```
Ditto comment about repeating "predictive".

ToucheSir commented on docs/src/models/overview.md:
```diff
@@ -159,36 +127,35 @@ julia> for epoch in 1:200
-After we trained the model, we verified it with the test data to verify the results. 
+1. Setup training and test data.
+2. Build a predictive model with initialized parameters.
+3. Find better parameters by optimizing a loss function that penalizes poor predictions.
```
```suggestion
3. Find better parameters by optimizing a loss function.
```
Something about the addition of the last 3 words doesn't sit 100% with me, but hard-pressed to come up with an alternative. Going to appeal to the wisdom of the crowd here.

ToucheSir reviewed (changes requested):
Thanks for the PR! I've just done a first pass review.

mattiasvillani committed:
Update docs/src/models/overview.md

Co-authored-by: Brian Chen <ToucheSir@users.noreply.github.com>

mattiasvillani committed:
Update docs/src/models/overview.md

Co-authored-by: Brian Chen <ToucheSir@users.noreply.github.com>

mattiasvillani committed:
Update docs/src/models/overview.md

Co-authored-by: Brian Chen <ToucheSir@users.noreply.github.com>

mattiasvillani committed:
Update docs/src/models/overview.md

Co-authored-by: Brian Chen <ToucheSir@users.noreply.github.com>

mattiasvillani committed:
Update docs/src/models/overview.md

Co-authored-by: Brian Chen <ToucheSir@users.noreply.github.com>

mattiasvillani committed:
Update docs/src/models/overview.md

Co-authored-by: Brian Chen <ToucheSir@users.noreply.github.com>

mattiasvillani committed:
Update docs/src/models/overview.md

Co-authored-by: Brian Chen <ToucheSir@users.noreply.github.com>

mattiasvillani committed:
Update docs/src/models/overview.md

Co-authored-by: Brian Chen <ToucheSir@users.noreply.github.com>

mattiasvillani committed:
Update docs/src/models/overview.md

Co-authored-by: Brian Chen <ToucheSir@users.noreply.github.com>

DhairyaLGandhi commented on docs/src/models/overview.md:
```diff
@@ -79,7 +79,7 @@ julia> loss(x_train, y_train)
 More accurate predictions will yield a lower loss. This loss function is called [mean squared error](https://www.statisticshowto.com/probability-and-statistics/statistics-definitions/mean-squared-error/). You can write your own loss functions or rely on those already provided by Flux. Flux works by iteratively reducing the loss through *training*.
 
-The Flux [`train!`](@ref) function uses *a loss function* and *training data* to improve the *parameters* of your model based on an [`optimiser`](../training/optimisers.md). We use the gradient `Descent` optimiser:
+The Flux [`train!`](@ref) function uses *a loss function* and *training data* to improve the *parameters* of your model based on an [`optimiser`](../training/optimisers.md). Here, we use basic gradient descent:
```
We have the `Descent` in there to train readers on our API. Repeating these several times was deliberate, but we used to have a much more terse doc style previously.

mattiasvillani replied:
Yes, I had it in my PR, but then accepted a suggested change from @ToucheSir . I would be happy to but `Descent` back.

darsnack replied:
```suggestion
The Flux [`train!`](@ref) function uses *a loss function* and *training data* to improve the *parameters* of your model based on an [`optimiser`](../training/optimisers.md). Here, we use `Descent`, a basic gradient descent optimizer, to improve our predictions:
```

ToucheSir replied:
I figured `Descent` was mentioned in the adjacent code block, but if it's being added back in we should take the opportunity to `@ref` it.

darsnack commented on docs/src/models/overview.md:
```diff
@@ -4,21 +4,19 @@ Flux is a pure Julia ML stack that allows you to build predictive models. Here a
-## Make a Trivial Prediction
-
-This example will predict the output of the function `4x + 2`. First, import `Flux` and define the function we want to simulate:
+This example will predict the output of the simple function `4x + 2`. First, import `Flux` and define the function we want to learn from training data:
```
I don't think we should say "training data" here because it's still just a function at this stage (we don't create any training data in this code block).

```suggestion
This example will predict the output of the simple function `4x + 2`. First, import `Flux` and define the function we want to learn:
```

darsnack commented on docs/src/models/overview.md:
````diff
@@ -48,44 +46,40 @@ Dense(1, 1)
-```julia
-julia> predict = Dense(1, 1)
-```
+Under the hood, a `Dense` layer is a struct with fields `weight` and `bias`. `weight` is a matrix of model weights and `bias` represents a bias vector. By default, these model parameters are initialized randomly.
````
The goal in the original writing of this tutorial was specifically to avoid "models" and to treat the model like a function that makes predictions. Though I agree that introducing `model` above then `predict` below is confusing. How about something like this:

````
Now, let's build a model to make predictions with `1` input and `1` output. In Flux, *models are conceptually predictive functions*:

```julia
julia> predict = Dense(1, 1)
Dense(1, 1)

julia> model.weight
1×1 Matrix{Float32}:
 -0.067352235

julia> model.bias
1-element Vector{Float32}:
 0.0
```

Under the hood, a `Dense` layer is a struct with fields `weight` and `bias`. `weight` is a matrix of model weights and `bias` represents a bias vector. By default, these model parameters are initialized randomly.

`Dense(1, 1)` implements the function `σ(Wx+b)` where `W` and `b` are the weights and biases. `σ` is an activation function that applies a (possibly nonlinear) transformation. Our model has one weight and one bias, but typical models will have many more. Think of weights and biases as knobs and levers Flux can use to tune predictions.
````

And we'd change `model` back to `predict` below.

ToucheSir replied:
Perhaps we should just add a sentence to that effect? Models are callable functions, all you need to do to get predictions is call the model on inputs, etc.

darsnack replied:
Maybe "In Flux, *models are conceptually predictive functions*" ➡️ "In Flux, *models are conceptually predictive functions* (i.e. we call our model to make predictions):"

mattiasvillani replied:
I like the comment about mentioning how a model is called to make predictions. That would be good to add. But I don't agree with using predict instead of model. See for example how keras in R and Python both uses model:

https://tensorflow.rstudio.com/guide/keras/
https://keras.io/examples/vision/image_classification_from_scratch/

darsnack commented on docs/src/models/overview.md:
```diff
@@ -48,44 +46,40 @@ Dense(1, 1)
+`Dense(1, 1)` implements the function `σ(Wx+b)` where `W` and `b` are the weights and biases. `σ` is an activation function that applies a (possibly nonlinear) transformation. Our model has one weight and one bias, but typical models will have many more. Think of weights and biases as knobs and levers Flux can use to tune predictions.
 
-This model will already make predictions, though not accurate ones yet:
+This model will already make predictions, though not accurate ones since we have not trained the model yet:
```
```suggestion
This model will already make predictions, though not accurate ones, since we have not trained the model yet:
```

darsnack commented on docs/src/models/overview.md:
```diff
@@ -98,59 +92,33 @@ julia> data = [(x_train, y_train)]
-
-The first parameter is the weight and the second is the bias. Flux will adjust predictions by iteratively changing these parameters according to the optimizer.
-
-This optimiser implements the classic gradient descent strategy. Now improve the parameters of the model with a call to [`train!`](@ref) like this:
+The [`train!`](@ref) function in Flux will use the optimizer to iteratively adjust the weights and biases in `parameters` to minimize the loss function. Calling `train!` once makes the optimizer take a single step toward the minimum: 
```
Following up on [@ToucheSir's comment](https://github.com/FluxML/Flux.jl/pull/1687/files#r681260203), I think these deletions should be reverted. The goal with this tutorial was to be deliberately slow and repetitive. I certainly agree that there is some fine-tuning of the wording that can be done in this section, but a complete deletion is not what we want.

darsnack commented on docs/src/models/overview.md:
````diff
@@ -98,59 +92,33 @@ julia> data = [(x_train, y_train)]
 ```
 
-It went down. Why? 
+And the parameters were changed by the optimizer:
````
Here I would revert back to the original intent: the reason the loss went down is because the parameters changed. We want to capture that causal relationship.

darsnack commented on docs/src/models/overview.md:
```diff
@@ -159,36 +127,35 @@ julia> for epoch in 1:200
+1. Setup training and test data.
+2. Build a model with initialized parameters.
+3. Find better parameters by optimizing a loss function.
+4. Evaluate the learned model by assessing performance on test data.
```
I would keep this bulleted list.

darsnack commented on docs/src/models/overview.md:
```diff
@@ -159,36 +127,35 @@ julia> for epoch in 1:200
+As expected, the predictions are very accurate since the underlying function was simple and observed without noise. Later on we will see `Flux` in action on substantially more challenging applications.
 
-After building the model, we trained it with `train!(loss, parameters, data, opt)`. The loss function is first, followed by the `parameters` holding the weights and biases of the model, the training data, and the `Descent` optimizer provided by Flux. We ran the training step once, and observed that the parameters changed and the loss went down. Then, we ran the `train!` many times to finish the t ...
+In summary, building and evaluating a predictive model in Flux follows the following steps:
```
```suggestion
In order to improve our predictions, we trained the model with `Flux.train!` and the `Descent` optimiser. We ran one training step and observed that the parameters changed and the loss went down. Finally, we repeated this training step many times to complete the learning process.

In summary, building and evaluating a predictive model in Flux follows the following steps:
```

darsnack reviewed (changes requested):
Thanks, I think this PR is doing a lot of nice changes to improve clarity. My only major requests are to re-include some of the longer paragraphs. They are intentionally long and deliberate to orient absolute beginners.

Of course, this style of tutorial should be complemented by more advanced tutorials. I think some of the work here could be incorporated into building a separate tutorial that orients more intermediate users. We could even add a disclaimer at the start of this tutorial to say "if you are already familiar with machine learning, you might prefer to start at X."

darsnack commented:
Might also be helpful to skim #1579 to see some of the motivation behind the original tutorial.

mattiasvillani commented:
I have revised the PR now to reflect all the nice input from @ToucheSir  @darsnack and @DhairyaLGandhi. I am new to opening up PRs on Docs and to have comments like this (but I have 20+ years of teaching experience as professor, partly in ML), and I am unsure on how to proceed. Perhaps easier to close this PR and open a new one with revised version? Please advice.

darsnack commented:
How did you create this PR? Did you use the GitHub web interface?

mattiasvillani commented:
Yes. But my current version is locally on the fork created by the GitHub
web version.

On Wed, Aug 4, 2021, 00:09 Kyle Daruwalla ***@***.***> wrote:

> How did you create this PR? Did you use the GitHub web interface?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/FluxML/Flux.jl/pull/1687#issuecomment-892199365>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJ2SHZ6XYWUTPNKKD2TYFTT3BSLFANCNFSM5BNNWZRQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>

darsnack commented:
You can continue to use the Github web interface to edit files. Just make sure you are editing your local fork and the `patch-1` branch (link here: https://github.com/mattiasvillani/Flux.jl/tree/patch-1). Edit any file and when you want to save the changes, you make a commit directly to `patch-1`. The PR will automatically incorporate those new commits.

As for dealing with the comments, I would just refer to the comments and make the changes you think are necessary manually by editing your fork's branch like described above. Don't bother trying to "commit the suggestions" in the PR web view.

mattiasvillani committed:
Update overview.md

This is the revised version after taking into account the comments from reviewers. See also my direct answers to reviewers regarding suggested changes that I did not (fully) implement.

mattiasvillani commented:
Ok, thanks. I just pushed to patch-1 in my forked repo, and the PR was
updated.
And thanks for the tip about dealing with comments. Makes sense and will
use that in the future.

Mattias

On Wed, Aug 4, 2021 at 12:34 AM Kyle Daruwalla ***@***.***>
wrote:

> You can continue to use the Github web interface to edit files. Just make
> sure you are editing your local fork and the patch-1 branch (link here:
> https://github.com/mattiasvillani/Flux.jl/tree/patch-1). Edit any file
> and when you want to save the changes, you make a commit directly to
> patch-1. The PR will automatically incorporate those new commits.
>
> As for dealing with the comments, I would just refer to the comments and
> make the changes you think are necessary manually by editing your fork's
> branch like described above. Don't bother trying to "commit the
> suggestions" in the PR web view.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/FluxML/Flux.jl/pull/1687#issuecomment-892210096>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJ2SH6KTS7DVTHLC5677HTT3BVGFANCNFSM5BNNWZRQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>


-- 
Mattias Villani, Professor

Department of Statistics
Stockholm University
e-mail: ***@***.***

Dept. of Computer and Information Science
Linköping University
e-mail: ***@***.***

Web: https://mattiasvillani.com <http://mattiasvillani.com/>

darsnack reviewed (changes requested):
I do like the new flow, especially the new titles. I think the verbosity of the summary should be somewhere between the original and what's in the current version. Certain paragraphs are probably too redundant, but I do think a few could be helpful (I'm primarily thinking of when I've taught students with no CS/ML/stats background).

mattiasvillani committed:
Another update after additional comments

I have changed the summary part at the end, based on suggestions from @darsnack (which I partially followed). Hopefully this is the final version.

mattiasvillani commented:
Thanks for the comments.
I commited a new revised version where the summary is modified, partially
following the suggestions from ***@***.****
Let me know if I need to do anything else for now.

Mattias

On Wed, Aug 4, 2021 at 1:03 AM Kyle Daruwalla ***@***.***>
wrote:

> ***@***.**** requested changes on this pull request.
>
> I do like the new flow, especially the new titles. I think the verbosity
> of the summary should be somewhere between the original and what's in the
> current version. Certain paragraphs are probably too redundant, but I do
> think a few could be helpful (I'm primarily thinking of when I've taught
> students with no CS/ML/stats background).
> ------------------------------
>
> In docs/src/models/overview.md
> <https://github.com/FluxML/Flux.jl/pull/1687#discussion_r682151602>:
>
> >  1×6 Matrix{Float32}:
>
>   0.0  -1.4925  -2.98501  -4.47751  -5.97001  -7.46252
>
>  ```
>
>
>
> -In order to make better predictions, you'll need to provide a *loss function* to tell Flux how to objectively *evaluate* the quality of a prediction. Loss functions compute the cumulative distance between actual values and predictions.
>
> +## Step 3 - Learn the model parameters on training data
>
> +
>
> +In order to make better predictions, you'll need to provide a *loss function* to tell Flux how to objectively *evaluate* the quality of a prediction. Loss functions compute the average distance between actual values and predictions so that more accurate predictions will yield a lower loss. Here is an example with the [mean squared error](https://www.statisticshowto.com/probability-and-statistics/statistics-definitions/mean-squared-error/) loss function:
>
>
> ⬇️ Suggested change
>
> -In order to make better predictions, you'll need to provide a *loss function* to tell Flux how to objectively *evaluate* the quality of a prediction. Loss functions compute the average distance between actual values and predictions so that more accurate predictions will yield a lower loss. Here is an example with the [mean squared error](https://www.statisticshowto.com/probability-and-statistics/statistics-definitions/mean-squared-error/) loss function:
>
> +In order to make better predictions, you'll need to provide a *loss function* to tell Flux how to objectively *evaluate* the quality of a prediction. Generally, loss functions compute the average distance between actual values and predictions so that more accurate predictions will yield a lower loss. Here is an example with the [mean squared error](https://www.statisticshowto.com/probability-and-statistics/statistics-definitions/mean-squared-error/) (MSE) loss function:
>
>
> ------------------------------
>
> In docs/src/models/overview.md
> <https://github.com/FluxML/Flux.jl/pull/1687#discussion_r682155106>:
>
> > @@ -181,14 +153,11 @@ julia> y_test
>
>   26  30  34  38  42
>
>  ```
>
>
>
> -The predictions are good. Here's how we got there.
>
> -
>
> -First, we gathered real-world data into the variables `x_train`, `y_train`, `x_test`, and `y_test`. The `x_*` data defines inputs, and the `y_*` data defines outputs. The `*_train` data is for training the model, and the `*_test` data is for verifying the model. Our data was based on the function `4x + 2`.
>
>
> I still feel that this particular paragraph can be helpful reinforcement
> for someone who has never seen machine learning. I'd imagine a student
> would read this and think "ok, I always need x_train, y_train, x_test, and
> y_test."
> ------------------------------
>
> In docs/src/models/overview.md
> <https://github.com/FluxML/Flux.jl/pull/1687#discussion_r682155995>:
>
> >  ```
>
>
>
> -The first parameter is the weight and the second is the bias. Flux will adjust predictions by iteratively changing these parameters according to the optimizer.
>
> -
>
> -This optimiser implements the classic gradient descent strategy. Now improve the parameters of the model with a call to ***@***.***) like this:
>
> +The ***@***.***) function in Flux will use the optimizer to iteratively adjust the weights and biases in `parameters` to minimize the loss function. Calling `train!` once makes the optimizer take a single step toward the minimum:
>
>
> ⬇️ Suggested change
>
> -The ***@***.***) function in Flux will use the optimizer to iteratively adjust the weights and biases in `parameters` to minimize the loss function. Calling `train!` once makes the optimizer take a single step toward the minimum:
>
> +The ***@***.***) function in Flux will use the optimizer to iteratively adjust the weights and biases in `parameters` to minimize the loss function. Calling `train!` once makes the optimizer take a single step toward the minimum of our loss function:
>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/FluxML/Flux.jl/pull/1687#pullrequestreview-721706007>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJ2SHZDCIE5UB3ZS3CZDS3T3BYV5ANCNFSM5BNNWZRQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>


-- 
Mattias Villani, Professor

Department of Statistics
Stockholm University
e-mail: ***@***.***

Dept. of Computer and Information Science
Linköping University
e-mail: ***@***.***

Web: https://mattiasvillani.com <http://mattiasvillani.com/>

mcabbott commented on docs/src/models/overview.md:
```diff
@@ -55,37 +48,33 @@ julia> model.bias
+Under the hood, a `Dense` layer is a struct with fields `weight` and `bias`. The field `weight` holds a matrix of model weights and `bias` represents a bias vector. By default, these model parameters are initialized randomly, and will later be learned from data by optimization. 
 
-`Dense(1, 1)` also implements the function `σ(Wx+b)` where `W` and `b` are the weights and biases. `σ` is an activation function (more on activations later). Our model has one weight and one bias, but typical models will have many more. Think of weights and biases as knobs and levers Flux can use to tune predictions. Activation functions are transformations that tailor models to your needs. 
+`Dense(1, 1)` also implements the function `σ(Wx+b)` where `W` and `b` are the weights and biases. `σ` is an activation function that applies a (possibly nonlinear) transformation. Our model has one weight and one bias, but typical models will have many more. Think of weights and biases as knobs and levers Flux can use to tune predictions.
```
There's some confusion here whether "one weight" means one scalar or one matrix. Both are true, of course, and both numbers would be larger. 
```suggestion
`Dense(1, 1)` also implements the function `σ(Wx+b)` on input `x`, where `W` and `b` are the weights and biases, and `σ` is an activation function that applies a (possibly nonlinear) transformation. In out model, `model.weight` is a 1×1 Matrix, just one number, and one `model.bias` a 1-element Vector. Typical models will have larger arrays, and more of them. Think of weights and biases as knobs and levers Flux can use to tune predictions.
```

mattiasvillani replied:
I implemented a version of this suggestion.

mattiasvillani commented on docs/src/models/overview.md:
```diff
@@ -1,48 +1,41 @@
-Under the hood, Flux uses a technique called automatic differentiation to take gradients that help improve predictions. Flux is also fully written in Julia so you can easily replace any layer of Flux with your own code to improve your understanding or satisfy special requirements.
+1. **Set up data** for training and test 
+2. **Build a model** with configurable parameters to make predictions
+3. **Learn the model parameters** on training data by iteratively updating the parameters to improve predictions
```
I also didn't like the repetion, but it my previous change was reverted by another reviewer. I have now implemented the second option since this first overview is supposed to be not too detailed.

mattiasvillani replied:
I took your text.

mattiasvillani commented on docs/src/models/overview.md:
````diff
@@ -55,37 +48,33 @@ julia> model.bias
-```julia
-julia> predict = Dense(1, 1)
-```
+Under the hood, a `Dense` layer is a struct with fields `weight` and `bias`. The field `weight` holds a matrix of model weights and `bias` represents a bias vector. By default, these model parameters are initialized randomly, and will later be learned from data by optimization. 
````
I added a less detailed version:

Notice that `x_train` is a 1×6 Matrix, not a vector, since it represents a batch of six different data points, rather than a single six-dimensional input.

mattiasvillani commented on docs/src/models/overview.md:
````diff
@@ -55,37 +48,33 @@ julia> model.bias
 ```julia
-julia> predict(x_train)
+julia> model(x_train)
 1×6 Matrix{Float32}:
  0.0  -1.4925  -2.98501  -4.47751  -5.97001  -7.46252
````
I added a less detailed version:

Notice that x_train is a 1×6 Matrix, not a vector, since it represents a batch of six different data points, rather than a single six-dimensional input.

mattiasvillani commented on docs/src/models/overview.md:
```diff
@@ -1,48 +1,41 @@
 
-Under the hood, Flux uses a technique called automatic differentiation to take gradients that help improve predictions. Flux is also fully written in Julia so you can easily replace any layer of Flux with your own code to improve your understanding or satisfy special requirements.
+1. **Set up data** for training and test 
+2. **Build a model** with configurable parameters to make predictions
```
To me, the loss function is part of Step 3, the learning. I also think that these four headings should be short with a minimum of details.

mattiasvillani commented on docs/src/models/overview.md:
```diff
@@ -98,60 +87,43 @@ julia> data = [(x_train, y_train)]
-The first parameter is the weight and the second is the bias. Flux will adjust predictions by iteratively changing these parameters according to the optimizer.
-
-This optimiser implements the classic gradient descent strategy. Now improve the parameters of the model with a call to [`train!`](@ref) like this:
+The [`train!`](@ref) function in Flux will use the optimizer to iteratively adjust the weights and biases in `parameters` to minimize the loss function. Calling `train!` once makes the optimizer take a single step toward the minimum of our loss function:
```
Implemented.

mattiasvillani commented on docs/src/models/overview.md:
```diff
@@ -98,60 +87,43 @@ julia> data = [(x_train, y_train)]
  ([0 1 … 4 5], [2 6 … 18 22])
```
I agree that it is complex object, but I also think that we don't want to go into too much detail in this first tutorial. I think the reader may just read this as a way to pack the data for use in train!. A lot of the details on dimensions and data structures may be better to present in later tutorial where the example is less contrived with proper dimensions > 1.

mattiasvillani commented on docs/src/models/overview.md:
````diff
@@ -98,60 +87,43 @@ julia> data = [(x_train, y_train)]
-```
-julia> parameters = params(predict)
+julia> parameters = params(model)
 Params([[-0.99009055], [0.0]])
````
Not sure about what you mean here, sorry. Can you clarify?

mcabbott replied:
I guess it's not released yet, but in the interests of future-proofing... from Julia 1.7 the printing of arrays should always show their dimensionality, while up to 1.6 one-column matrices and vectors look identical:
```
julia> ones(1), ones(1,1), ones(1,1,1)
([1.0], [1.0;;], [1.0;;;])

julia> VERSION
v"1.7.0-beta3.0"

#

julia> ones(1), ones(1,1), ones(1,1,1)
([1.0], [1.0], [1.0])

julia> VERSION
v"1.6.0"
```
